File: Data Structures & Algorithms/count-subsequences/submission-2.java
class Solution {
    private int c=0;
    public void dfs(int i,int j , List<Character> cur , String s , String t){
        if(i>=s.length() || j>= t.length()) return ;
        if(s.charAt(i)==t.charAt(j)){
            cur.add(s.charAt(i));
            if(cur.size()==t.length()){
                c+=1;
            }
            else{
                dfs(i+1,j+1,cur,s,t);
            }
            cur.removeLast();
            dfs(i+1,j,cur,s,t);
        }
        else 
           dfs(i+1,j,cur,s,t);
    }
    public int numDistinct(String s, String t) {
        dfs(0,0,new ArrayList<>() , s,t);
        return c ;
    }
}

File: Data Structures & Algorithms/target-sum/submission-2.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        n=len(nums)
        d={}
        def dfs(i, cur):
            if (i,cur) in d:
                return d[(i,cur)]
            elif i==n-1 and cur != target:
                return 0
            elif i==n-1 and cur == target :
                return 1
            else:
                d[ (i,cur) ] = dfs(i+1, cur + nums[i+1]) + dfs(i+1, cur - nums[i+1])
                return d[(i,cur)]
        return dfs(0,-nums[0])+dfs(0,nums[0])